### PR TITLE
Added new parameter to ignore inject namespaces.

### PR DESCRIPTION
**Pull request by gyliu513**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes https://github.com/istio/pilot/issues/1433

**Special notes for your reviewer**:

**Release note**:

```release-note
Added new parameter of ignoredNamespaces to sidecar initializer configMap
```

/cc @linsun @ayj

### CODE REVIEW

File: platform/kube/inject/inject.go [hunk 1]
```diff
@@ -460,6 +463,29 @@ func intoObject(c *Config, in interface{}) (interface{}, error) {
...
+			if c.Namespaces[0] == "" {
+				glog.V(2).Infof("The namespace %s will be ignored by sidecar initializer", ignoredNamespace)
+			} else {
+				for _, namespace := range c.Namespaces {
+					if ignoredNamespace == namespace {
+						glog.V(2).Infof("The namespace %s will be ignored by sidecar initializer", ignoredNamespace)
+						break
+					}
+				}
+			}
```
Comment (linsun) [inline platform/kube/inject/inject.go]: Is it necessary to do the if/else block?  It seems you'll ignore the namespace as long as it is on the ignoredNamespace list, regardless if it is specified in c.Namespaces.

Comment (gyliu513) [reply platform/kube/inject/inject.go]: Here the logic is that I want to make the `ignoredNamespace` has high priority than `Namespaces`.

The if/else logic is mainly used to print logs telling end user that the `ignoredNamespace` is overwriting the `Namespaces`.

We can remove the if/else logic, but then we will not able to check if the `ignoredNamespace` is overwriting the `Namespaces` or not.

Comment (linsun) [reply platform/kube/inject/inject.go]: I see your point now.  If it is mainly for the purpose of logging, you may want to clarify in that log that you are overwriting the `Namespaces`.

File: platform/kube/inject/inject.go [hunk 2]
```diff
@@ -161,6 +161,9 @@ type Config struct {
 	// deprecate if InitializerConfiguration becomes namespace aware
 	Namespaces []string `json:"namespaces"`
```
Comment (rshriram) [inline platform/kube/inject/inject.go]: rename variable to IncludeNamespaces and the one below to ExcludeNamespaces.
Ideally, we should rename the `json:"namespaces"` to `json:"includeNamespaces"` as well.

File: platform/kube/inject/inject.go [hunk 3]
```diff
@@ -460,6 +463,29 @@ func intoObject(c *Config, in interface{}) (interface{}, error) {
...
+		if obj.GetNamespace() == ignoredNamespace {
+			// If the namesepace was defined in both IgnoredNamespaces and Namespaces,
+			// then IgnoredNamespaces will overwrite the Namespaces.
+			if c.Namespaces[0] == "" {
+				glog.V(2).Infof("The namespace %s will be ignored by sidecar initializer", ignoredNamespace)
+			} else {
+				for _, namespace := range c.Namespaces {
+					if ignoredNamespace == namespace {
+						glog.V(2).Infof("The namespace %s will be ignored by sidecar initializer", ignoredNamespace)
+						break
```
Comment (rshriram) [inline platform/kube/inject/inject.go]: This is unnecessary and will cause confusion. Its better to have a mutually exclusive model.
User can specify either IncludeNamespaces or ExcludeNamespaces, but not both, where the semantics are
if ExcludeNamespaces is provided, sidecars will be injected on ALL except the namespaces provided..
if IncludeNamespaces is provided, sidecars will be injected on ONLY the namespaces provided.

File: platform/kube/inject/inject_test.go [hunk 4]
```diff
@@ -152,6 +152,14 @@ func TestIntoResourceFile(t *testing.T) {
...
 			want: "testdata/replicationcontroller.yaml.injected",
 		},
+		{
+			in:   "testdata/hello-ibm.yaml",
+			want: "testdata/hello-ibm.yaml.inject",
+		},
+		{
+			in:   "testdata/hello-google.yaml",
+			want: "testdata/hello-google.yaml.inject",
+		},
```
Comment (rshriram) [inline platform/kube/inject/inject_test.go]: why do you need two ? They both seem to be excluding ibm-system namespace. Might as well just have one.

File: platform/kube/inject/inject.go (hunk 1 above)
Comment (gyliu513) [reply platform/kube/inject/inject.go]: Sure, let's wait some comments from @ayj  ;-)

File: platform/kube/inject/inject.go (hunk 2 above)
Comment (linsun) [reply platform/kube/inject/inject.go]: I'm concerned on the renaming...  our docs and 0.2.7 is out, I'd prefer not rename the existing `namespaces` support in a point release. I also agree with @gyliu513 that we need to support namespaces: [""] and a few excluded namespace.  This design was discussed in details in https://github.com/istio/istio.github.io/issues/648

Comment (rshriram) [reply platform/kube/inject/inject.go]: You don't have to necessarily rename the json field. Just rename the variables internally at the very least. And see my comment below.. the excludedNamespaces option supports the use case you mentioned.

File: platform/kube/inject/inject.go (hunk 3 above)
Comment (gyliu513) [reply platform/kube/inject/inject.go]: Done

File: platform/kube/inject/inject_test.go (hunk 4 above)
Comment (gyliu513) [reply platform/kube/inject/inject_test.go]: My original thinking is that i want to test the cases where the excludeNamespaces have multiple values, but yes, have one value is good enough.

File: platform/kube/inject/inject.go (hunk 2 above)
Comment (gyliu513) [reply platform/kube/inject/inject.go]: OK, I created another patch here https://github.com/istio/pilot/pull/1437 , and will close this one.

**Comment by gyliu513** [comment]
Please check the new PR here https://github.com/istio/pilot/pull/1437 

/cc @linsun @rshriram @ayj

**Pull request closed without merging**